typing/ctype.ml

```diff
@@ -2568,6 +2569,24 @@ let unify1_var env t1 t2 =
     Private_type_expr.set_desc t1 d1;
     raise e
 
+(* Do not use local constraints more than necessary *)
+let rec unify_public env t1 t2 =
+  let t1 = repr t1 and t2 = repr t2 in
+  if unify_eq t1 t2 then () else
+  match (t1.desc, t2.desc) with
+  | (Tconstr (p1, tl1, _), Tconstr (p2, tl2, _))
+    when is_public_type env p1 && is_public_type env p2 ->
```

Comment:
Do we actually need this `when` clause? It's not clear to me why the type being private or being a variant/record would make the actions in this clause unsound.

Reply:
I shall probably write a more explicit comment.
Contrary to `unify2`, this code forgets the abbreviations it is expanding.
In particular this means that since `unify2` should be called on the non-expanded types, we cannot call it directly here, hence the exception `Cannot_expand` when reverting to `unify2`.
Since we do not want to do useless work, and the case which interests us is when there is a risk of superfluously expanding a GADT equation, we stop as soon as there is no chance to reach one.
Since GADT equations are only defined for abstract types, and since an abstract type can only be reached by expanding a _public_ type, it is sufficient to consider public types.